Added Continuous TF Collection

Kazadhum opened the pull request:
Hi @miguelriemoliveira, @manuelgitgomes and @rarrais! 

Following the implementation of continuous sensor data collection in #1010, this PR addresses the implementation of continuous TF and joint data collection, described in issue #1014.

Relatively few changes were made overall.

I added a new parameter in the calibration configuration files:

https://github.com/lardemua/atom/blob/b1091f44101d4d3ba67b157ee20f11cb59ae7f85/atom_calibration/templates/config.yml#L112-L113

When set to True, TF data and joint data will be saved continuously. To ensure backwards compatibility, it is assumed to be False if its not specified in the config file.

The structure of the continuous data was also changed from #1010. It now looks like this:

![image](https://github.com/user-attachments/assets/2b9aa42d-d649-44cd-adf2-b9e70eaa0e58)

Can this be merged?

Kazadhum committed:
chore(rihibot): readded to atom_examples after pr #1010

Kazadhum committed:
chore(liang): began implementation of rk4 integrator for liang method

Kazadhum committed:
chore(liang): separated integator script from liang #1012

I've separated the integrator funtion from the Liang calibration method
script for better modularity and easier testing.

Kazadhum committed:
chore(rihibot): re-config after renaming imu sensor #1012

Kazadhum committed:
feat(integrator): initial implementation of integrator #1012

Kazadhum committed:
test(integrator): testing integrator for potential fixes #1012

Kazadhum committed:
feat(integrator): added euler method for integration of orientation #1012

Added the Euler method for integrating the orientation of the IMU.
Weirdly, this simpler and theoretically worse algorithm is outputting
errors much smaller than RK4. A bug in the implementation of RK4 is
probably involved. Investigating...

Kazadhum committed:
minor(rotations): added a matrixToTranslationQuaternion function

Kazadhum committed:
minor(ci): added unit tests for rk4

Kazadhum committed:
chore: debugging rk4

Kazadhum committed:
feat(config): config.yml now accepts a continuous_joint_and_tf_collection option #1014

Added support for a 'continuous_joint_and_tf_collection' flag in the
config.yml;
Maintained backwards compatibility by only issuing a warning if that key
is missing.

Kazadhum committed:
feat(data_collection): added continuous tf collection #1014

Added a mechanism for saving every TF message during data collection;
Needs some adjustments, namely:
(1) the addition of a flag to opt into this feature -- this flag should come form the config.yml file (the changes
to the configuration script have already been made);
(2) the addition of a time stamp with each TF message saved (when saving
tfs in a regular collection this isn't needed, as that timestamp is used
for the collection and not for a singular TF message, hance the
adjustments needed);
(3) the addition of a continuous joint states collection, to complement
the TFs, which will be needed for some systems.

Kazadhum committed:
chore(data_collection): added opt-in with config key #1014

Now only continuously saves TFs if the
'continuous_joint_and_tf_collection' key in config.yml is set to True

Kazadhum committed:
chore(rihibot): added continuous_joint_and_tf_collection flag in config #1014

Kazadhum committed:
chore(rihibot): added timestamp for each TF in continuous_tf_data #1014

Kazadhum committed:
feat(data_collection): added continuous joint state collection along with tfs #1014

Added the continuous collection of joint parameters along with TFs.

Kazadhum committed:
fix(data_collection): restructured continuous data in the dataset #1014

Restructured the way continuous data is saved in the dataset so it's all
in the same folder.

Kazadhum committed:
chore(imu_integrator): temporarily removed integrator for PR #1016

Kazadhum committed:
chore(rihibot): temporarily removed rihibot for PR #1016

miguelriemoliveira commented:
Hey @Kazadhum ,

thanks for the PR. 

> I added a new parameter in the calibration configuration files:
> 
> [atom/atom_calibration/templates/config.yml](https://github.com/lardemua/atom/blob/b1091f44101d4d3ba67b157ee20f11cb59ae7f85/atom_calibration/templates/config.yml#L112-L113)
> 
> Lines 112 to 113 in [b1091f4](https://github.com/lardemua/atom/commit/b1091f44101d4d3ba67b157ee20f11cb59ae7f85)
> 
>  # Enable continuous joint and transform collection (set to True or False) 
>  continuous_joint_and_tf_collection: False

now that we have the idea to have a **continuous_data** field which is general, meaning we can store not only continuous tf and joint data but also other topcis, would it not be better to have a way to specify which topics we want to have recorded in continuous mode.

Something like:

```yaml
continuous_data:
    topics: ['tf',  'joint_states']#a list of topics to record in continuous mode
    max_frequency: 20 # save a maximum of 20 Hz for each topic (this would required setting the message throtlers like in the sensors throtle field).
```

I noticed also that we can set other topics to continous mode in the sensors our additional_tfs field? Perhaps these could be removed and the list would control everything? What do you think?

Kazadhum commented:
Hi @miguelriemoliveira!

> now that we have the idea to have a continuous_data field which is general, meaning we can store not only continuous tf and joint data but also other topcis, would it not be better to have a way to specify which topics we want to have recorded in continuous mode.

I think it's a good idea to put everything related to continuous data collection in the same place, but I would still separate sensor and tf data/topics from the others. This is because what is continuously saved in the case of, for example, an RGB camera, is the labeled message topic. On the other hand, saving tfs implies unifying the /tf and /tf_static topics. These differences in the logic of data collection for each "type" of data lead me to believe it might be better to specify whether we're dealing with a sensor or tfs or another generic topic.

Perhaps this alternative:

```
continuous_data:
   sensors: ["sensor1", "sensor2"]
   tf: true/false
   joints: true/false
   other_topics/additional_topics: ["/topic1", "/topic2"]
```

With this, we differentiate the sensor data from the rest but still centralize the configuration for continuous data.

> I noticed also that we can set other topics to continous mode in the sensors our additional_tfs field? Perhaps these could be removed and the list would control everything? What do you think?

At the moment, there is no continuous data collection setting or field in the additional_tfs field. Either way, I think this would be solved with the solution I propose.

miguelriemoliveira commented:
Hey @Kazadhum ,

thanks for the reply. Not sure about the part 

> I would still separate sensor and tf data/topics from the others. This is because what is continuously saved in the case of, for example, an RGB camera, is the labeled message topic. 



> On the other hand, saving tfs implies unifying the /tf and /tf_static topics. These differences in the logic of data collection for each "type" of data lead me to believe it might be better to specify whether we're dealing with a sensor or tfs or another generic topic.


Why can't we save /tf and /tf_static topics continuously as well? I may be missing the point, but if we were able to not have an exception for a specific msg type that would be desirable.

Kazadhum commented:
Hi @miguelriemoliveira! Perhaps we could meet to discuss this more clearly, but I think it comes down to what we want the continuous data collection to output.

The regular data collection process, for each collection, saves a `transforms` dictionary with every transformation, collected from both the /tf and /tf_static messages. 

I was trying to recreate this exact behaviour, except continuously (i.e., for every message received in /tf, create that same dictionary with *every* transformation, including those from /tf_static). I think this might make it easier to use the data in future processes.

I think what you're suggesting, if I'm not mistaken, is that we simply save the ROS messages corresponding to a given list of topics to save. 

My worry is that it would make the posterior data usage slightly more complicated. In the case of the tfs, this union of the ones in /tf and the ones in /tf_static would have to be done after data collection, in whichever scripts use that data.

On the other hand, it does make data collection more generic, so it might be the way to go.

What do you think?

miguelriemoliveira commented:
> Hi @miguelriemoliveira! Perhaps we could meet to discuss this more clearly, but I think it comes down to what we want the continuous data collection to output.
> 
> The regular data collection process, for each collection, saves a `transforms` dictionary with every transformation, collected from both the /tf and /tf_static messages.
> 
> I was trying to recreate this exact behaviour, except continuously (i.e., for every message received in /tf, create that same dictionary with _every_ transformation, including those from /tf_static). I think this might make it easier to use the data in future processes.
> 
> I think what you're suggesting, if I'm not mistaken, is that we simply save the ROS messages corresponding to a given list of topics to save.
> 
> My worry is that it would make the posterior data usage slightly more complicated. In the case of the tfs, this union of the ones in /tf and the ones in /tf_static would have to be done after data collection, in whichever scripts use that data.
> 
> On the other hand, it does make data collection more generic, so it might be the way to go.
> 
> What do you think?

Hey @Kazadhum ,

yes, perhaps we can talk. Call me tomorrow please. I have a meeting at 11am, so it's fine before or after.

Kazadhum commented:
Hi @miguelriemoliveira ! Will do!

Kazadhum committed:
minor(data_collector): changed how continuous data collection works #1016

Changed the continuous data collection mechanism to be more generic.
Now, it receives a list of topics to continuously save (in the config
yaml) and simply saves every message in those topics.

Kazadhum commented:
Hi @miguelriemoliveira! I changed the mechanism according to what you suggested. We now have, in the configuration file:

https://github.com/lardemua/atom/blob/2178fe2d686fe1a633db502df1882d0bd67f18ee/atom_calibration/templates/config.yml#L108-L109

This creates one subscriber for each topic that saves every message received.

As an example, I have the following in the configuration of `rihibot`:
```
# Enable continuous data collection for a list of ROS message topics
continuous_data: ["/tf", "/tf_static", "/imu"]
```

Which leads to the following in the dataset:
![image](https://github.com/user-attachments/assets/5732d190-8b28-4375-8fa8-3e6c4a26c285)

No processing is done at this stage, so it is more generic.

What do you think @miguelriemoliveira?

miguelriemoliveira commented:
Hey @Kazadhum ,

it looks great from my side. I suggest that you go ahead and accept the PR. 

One comment before you do: in the template 

[atom/atom_calibration/templates/config.yml](https://github.com/lardemua/atom/blob/2178fe2d686fe1a633db502df1882d0bd67f18ee/atom_calibration/templates/config.yml#L108-L109)

the continous_data should have the topics commented since it is not a very often used option.

Kazadhum commented:
Hi @miguelriemoliveira! Thank you. I'll make that change and merge the PR. I will also delete the temporary `imu-calibration-2` branch to keep working on the `imu-calibration` branch.

Kazadhum committed:
chore(config_template): commented the 'continuous_data' field as its use is rare #1016
